misc/vtk/src/processors/vtktovolume.cpp

```diff
@@ -0,0 +1,422 @@
+        if (format->getId() != selectedFormat) continue;
+        int numComps = array->GetNumberOfComponents();
+
+        numTotalComps += numComps;
+        dataArrays_.addArray(arrayName, numComps, array->GetVoidPointer(0));
+        if (keepSettings) {
+            BoolProperty *prop = dynamic_cast<BoolProperty *>(dataArrays_[dataArrays_.size() - 1]);
+            if (prop && formerChannels.find(prop->getDisplayName()) != formerChannels.end())
+                prop->set(true);
+        }
+    }
+    if (!keepSettings && numTotalComps <= 4)
+        for (auto *prop : dataArrays_.getProperties())
+            if (BoolProperty *boolProp = dynamic_cast<BoolProperty *>(prop)) boolProp->set(true);
+
+    dataArrays_.selectedFormat = dataFormats_.get();
+
+    if (keepSettings) convertData_ = true;
+
+    reactToChange_ = true;
+}
+
+void VTKtoVolume::updateAvailableArrays() {
+    NetworkLock lock;
+    reactToChange_ = false;
+
+    int numTotalComps = 0;
+    auto properties = dataArrays_.getProperties();
+    for (size_t p = 0; p < properties.size(); ++p) {
+        if (dynamic_cast<BoolProperty *>(properties[p])->get())
+            numTotalComps += dataArrays_.numArrayComponents[p];
+    }
+
+    for (size_t p = 0; p < properties.size(); ++p) {
+        if (dynamic_cast<BoolProperty *>(properties[p])->get()) continue;
+        bool tooMuch = numTotalComps + dataArrays_.numArrayComponents[p] > 4;
+        properties[p]->setReadOnly(tooMuch);
+    }
+
+    reactToChange_ = true;
+}
+
+void VTKtoVolume::convertData() {
+    outport_.setData(nullptr);
+    if (!inport_.hasData()) return;
+
+    auto vtkDataPtr = inport_.getData();
+
+    // Get dimensions. If none are given, it's not a regular grid.
+    std::optional<size3_t> dimensionsOpt = vtkDataPtr->getDimensions();
+    if (!dimensionsOpt) {
+        LogWarn("Input vtk data set is not rectilinear.");
+        return;
+    }
+    size3_t dimensions = *dimensionsOpt;
+    if (useCellData_.get())
+        for (size_t d = 0; d < 3; ++d) dimensions[d]--;
+
+    struct RawArrayData {
+        unsigned char *ptr;
+        size_t byteSize;
+    };
+    std::vector<RawArrayData> arrayDataPointers;
+
+    // How many components in total?
+    const DataFormatBase *format = DataFormatBase::get(dataFormats_.getSelectedValue());
+    size_t sizeElement = format->getSize();
+    size_t numTotalComps = 0;
+    for (size_t p = 0; p < dataArrays_.size(); ++p) {
+        BoolProperty *prop = dynamic_cast<BoolProperty *>(dataArrays_[p]);
+        if (prop && prop->get()) {
+            arrayDataPointers.push_back(
+                {dataArrays_.ptrArrayData[p], dataArrays_.numArrayComponents[p] * sizeElement});
+            numTotalComps += dataArrays_.numArrayComponents[p];
+        }
+    }
+    if (!numTotalComps) {
+        LogWarn("No components selected.");
+        return;
+    }
+
+    const DataFormatBase *multichannelFormat =
```

Comment:
We might want to extract this to a vtk volume reader so that one can use the regular VolumeSource with vtk-data

Reply:
That can be tricky, since the vtk format is a bit weird. The old format `.vtk` supports everything. The newer one has different extensions per type of data, e.g. `.vti` for image, `.vtp` for points... This is taken care of by the generic VTK Source processor which only forwards a VTK data object.

Reply:
It could just throw an exception if failed to read/unsupported

Reply:
There are processors that can work with that kind of data, even if it's not on a  regular grid. So there is definitely a need for the VTK reader and converter in its current form.

Reply:
I agree, but think that it could be valid to have both. A later problem though.